File: src/dose/dose_java.ss/src/com.saabgroup.safir.dob/src/com/saabgroup/safir/dob/ConnectionAspectMisc.java
// -*- coding: utf-8 -*-
/******************************************************************************
*
* Copyright Saab AB, 2008-2013, 2022-2023 (http://safirsdkcore.com)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/

package com.saabgroup.safir.dob;
/**
 * Class that provides miscellaneous methods that are used less frequently.
 *
 */
public class ConnectionAspectMisc
    extends ConnectionAspectBase
{
    /**
     * Constructor
     *
     * @param connection The connection that you want to operate through.
     */
    public ConnectionAspectMisc(ConnectionBase connection) {
        super(connection);
    }


    /**
     * Get info about which callback you are currently executing in.
     *
     * @return Id of the callback you are currently inside, or None if not in a callback.
     */
    public CallbackId getCurrentCallbackId()
    {
        boolean [] success = new boolean [1];
        int [] callbackId = new int [1];
        Interface.GetCurrentCallbackId(getControllerId(), callbackId, success);

        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }

        return CallbackId.values()[callbackId[0]];
    }

    //
    // Connection info
    //


    /**
     * Get the name for this connection used in the system.
     *
     * The connection name is composed of the name parts given by the application
     * when opening the connection, with some additional decoration made by the DOB.
     *
     * @return The connection name.
     */
    public String getConnectionName()
    {
        boolean [] success = new boolean [1];
        String [] name = new String [1];
        Interface.GetConnectionName(getControllerId(), name, success);

        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }

        return name[0];

    }

    /**
     * Get the common part of the connection name.
     *
     * @return The connection name common part specified when opening the connection.
     */
    public String getConnectionNameCommonPart()
    {
        boolean [] success = new boolean [1];
        String [] name = new String[1];
        Interface.GetConnectionNameCommonPart(getControllerId(), name, success);

        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }

        return name[0];
    }

    /**
     * Get the instance part of the connection name.
     *
     * @return The connection name instance part specified when opening the connection.
     */
    public String getConnectionNameInstancePart()
    {
        boolean [] success = new boolean [1];
        String [] name = new String[1];
        Interface.GetConnectionNameInstancePart(getControllerId(), name, success);

        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }
        return name[0];
    }

    /**
     * Get the context that the connection is opened in.
     *
     * @return The context of the connection
     */
    public int getContext()
    {
        int [] context = new int[1];
        boolean [] success = new boolean [1];
        Interface.GetContext(getControllerId(), context, success);
        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }
        return context[0];
    }

    /**
     * Get the Node Identifier of the current node.
     *
     * Be aware that this identifier changes every time the node restarts.
     *
     * @return NodeId of current node
     */
    public long getNodeId()
    {
        long [] nodeId = new long[1];
        boolean [] success = new boolean [1];
        Interface.GetNodeId(nodeId, success);
        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }
        return nodeId[0];
    }


    /**
     * Check whether the current node is a light node.
     *
     * This is slightly easier than looking through the parameters to work this out.
     *
     * @return true if the current node is a light node.
     */
    public boolean isLightNode()
    {
        boolean [] isLight = new boolean[1];
        boolean [] success = new boolean [1];
        Interface.IsLightNode(getControllerId(), isLight, success);
        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }
        return isLight[0];
    }

    //
    // Queue Status
    //


    /**
     * Get the capacity of the specified queue.
     *
     * This method returns the maximum number of items that the queue can hold.
     *
     * @param queue The queue to get info for.
     * @return The capacity of the queue.
     */
    public int getQueueCapacity(com.saabgroup.safir.dob.ConnectionQueueId queue)
    {
        int [] queueCapacity = new int[1];
        boolean [] success = new boolean [1];
        Interface.GetQueueCapacity(getControllerId(), queue.ordinal(), queueCapacity, success);
        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }
        return queueCapacity[0];
    }

    /**
     * Get the number of items currently in the queue.
     *
     * This method returns the number of items that is currently in the specified queue.
     * NOTE: This method is only implemented for out-queues (MessageOutQueue, RequestOutQueue)
     *       If this method is called for an in-queue, a Safir::Dob::SoftwareViolationException 
     *       will be thrown.
     *
     * @param queue [in] The queue to get info for.
     * @return The current size of the queue.
     */
    public int getQueueSize(com.saabgroup.safir.dob.ConnectionQueueId queue)
    {
        int [] queueSize = new int[1];
        boolean [] success = new boolean [1];
        Interface.GetQueueSize(getControllerId(), queue.ordinal(), queueSize, success);
        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }
        return queueSize[0];
    }

    //
    // Debug
    //

    /**
     * Turn simulation of overflow on/off. For test purposes.
     *
     * Setting inQueues to true means that no messages or requests are handled by the application.
     * An incoming request will result in an overflow, and an incoming message will be discarded.
     * Setting outQueues to true means that no messages or requests can be sent from the application,
     * instead these calls will throw a Safir::Dob::OverflowException. When reset to false
     * OnXxxxNotOverflow will be called as expected.
     * Use this to verify that your application handles overflows correctly.
     *
     * Note that the inQueues flag is not applied to new consumers added after this call.
     *
     * @param inQueues If true all incoming queues are simulated full.
     * @param outQueues If true all outgoing queues are simulated full.
     */
    public void simulateOverflows(boolean inQueues, boolean outQueues)
    {
        boolean [] success = new boolean [1];
        Interface.SimulateOverflows(getControllerId(),
                                    inQueues,
                                    outQueues,
                                    success);
        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }
    }

    //
    // Shared Memory statistics
    //

    /**
     * Get the number of currently used bytes in the shared memory.
     *
     * The size of the shared memory is defined by the parameter
     * Safir.Dob.NodeParameters.SharedMemorySize, which is defined in megabytes (1024*1024 bytes).
     *
     * Calling this function does not require the underlying Connection to have been opened.
     *
     * @return The amount of shared memory used, in bytes.
     */
    public long getSharedMemoryUsage()
    {
        boolean [] success = new boolean [1];
        long [] usage = new long [1];
        Interface.GetSharedMemoryUsage(usage,
                                       success);
        if (!success[0]) {
            com.saabgroup.safir.dob.typesystem.LibraryExceptions.getInstance().Throw();
        }
        return usage[0];
    }
}

File: src/dose/dose_dotnet.ss/src/ConnectionAspectMisc.cs
/* ****************************************************************************
*
* Copyright Saab AB, 2008-2013, 2022 (http://safirsdkcore.com)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/

using System;

namespace Safir.Dob
{
    /// <summary>
    ///  Class that provides miscellaneous methods that are used less frequently.
    /// </summary>
    public sealed class ConnectionAspectMisc : ConnectionAspectBase
    {
        /// <summary>
        /// Get the current Safir instance, i.e reads the environment variable SAFIR_INSTANCE.
        /// </summary>
        public static int GetSafirInstance()
        {
            int instance = 0;
            var env = System.Environment.GetEnvironmentVariable("SAFIR_INSTANCE");
            if (env == null)
            {
                return 0;
            }
            if (int.TryParse(env, out instance))
            {
                return instance;
            }
            throw new System.Exception("SAFIR_INSTANCE is not set to a number");
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="connection">The connection that you want to operate through</param>
        public ConnectionAspectMisc(ConnectionBase connection) : base(connection) { }

        /// <summary>
        /// Get info about which callback you are currently executing in.
        /// </summary>
        /// <returns>Id of the callback you are currently inside, or None if not in a callback.</returns>
        public CallbackId.Enumeration GetCurrentCallbackId()
        {
            byte success;
            System.Int32 callbackId;
            Interface.DoseC_GetCurrentCallbackId(ControllerId, out callbackId, out success);

            if (!Interface.BoolOf(success))
            {
                Typesystem.LibraryExceptions.Instance.Throw();
            }
            return (Safir.Dob.CallbackId.Enumeration)callbackId;
        }

        #region Connection info
        /// <summary>
        /// Get the name for this connection used in the system.
        /// <para/>
        /// The connection name is composed of the name parts given by the application
        /// when opening the connection, with some additional decoration made by the DOB.
        /// </summary>
        /// <returns>The connection name.</returns>
        public string GetConnectionName()
        {
            byte success;
            System.IntPtr name;
            Interface.DoseC_GetConnectionName(ControllerId, out name, out success);

            if (!Interface.BoolOf(success))
            {
                Typesystem.LibraryExceptions.Instance.Throw();
            }
            return Typesystem.Internal.InternalOperations.StringOf(name);

        }

        /// <summary>
        /// Get the common part of the connection name.
        /// </summary>
        /// <returns>The connection name common part specified when opening the connection.</returns>
        public string GetConnectionNameCommonPart()
        {
            byte success;
            System.IntPtr name;
            Interface.DoseC_GetConnectionNameCommonPart(ControllerId, out name, out success);

            if (!Interface.BoolOf(success))
            {
                Typesystem.LibraryExceptions.Instance.Throw();
            }
            return Typesystem.Internal.InternalOperations.StringOf(name);
        }

        /// <summary>
        /// Get the instance part of the connection name.
        /// </summary>
        /// <returns>The connection name instance part specified when opening the connection.</returns>
        public string GetConnectionNameInstancePart()
        {
            byte success;
            System.IntPtr name;
            Interface.DoseC_GetConnectionNameInstancePart(ControllerId, out name, out success);

            if (!Interface.BoolOf(success))
            {
                Typesystem.LibraryExceptions.Instance.Throw();
            }
            return Typesystem.Internal.InternalOperations.StringOf(name);
        }

        /// <summary>
        /// Get the context that the connection is opened in.
        /// </summary>
        /// <returns>Context</returns>
        public System.Int32 GetContext()
        {
            System.Int32 context;
            byte success;
            Interface.DoseC_GetContext(ControllerId, out context, out success);
            if (!Interface.BoolOf(success))
            {
                Typesystem.LibraryExceptions.Instance.Throw();
            }
            return context;
        }

        /// <summary>
        /// Get the Node Identifier of the current node.
        /// <para/>
        /// Be aware that this identifier changes every time the node restarts.
        /// </summary>
        /// <returns>NodeId of current node</returns>
        public System.Int64 GetNodeId()
        {
            System.Int64 nodeId;
            byte success;
            Interface.DoseC_GetNodeId(out nodeId, out success);
            if (!Interface.BoolOf(success))
            {
                Typesystem.LibraryExceptions.Instance.Throw();
            }
            return nodeId;
        }
        #endregion

        #region Queue Status

        /// <summary>
        /// Get the capacity of the specified queue.
        /// <para/>
        /// This method returns the maximum number of items that the queue can hold.
        /// </summary>
        /// <param name="queue">The queue to get info for</param>
        /// <returns>The capacity of the queue</returns>
        public System.Int32 GetQueueCapacity(Safir.Dob.ConnectionQueueId.Enumeration queue)
        {
            System.Int32 queueCapacity;
            byte success;
            Interface.DoseC_GetQueueCapacity(ControllerId, (System.Int32)queue, out queueCapacity, out success);
            if (!Interface.BoolOf(success))
            {
                Typesystem.LibraryExceptions.Instance.Throw();
            }
            return queueCapacity;
        }

        /// <summary>
        /// Get the number of items currently in the queue.
        /// <para/>
        /// This method returns the number of items that is currently in the specified queue.
        /// NOTE: This method is only implemented for out-queues (MessageOutQueue, RequestOutQueue)
        ///       If this method is called for an in-queue, a SoftwareViolationException will be thrown.
        /// </summary>
        /// <param name="queue">The queue to get info for.</param>
        /// <returns>The current size of the queue.</returns>
        public System.Int32 GetQueueSize(Safir.Dob.ConnectionQueueId.Enumeration queue)
        {
            System.Int32 queueSize;
            byte success;
            Interface.DoseC_GetQueueSize(ControllerId, (System.Int32)queue, out queueSize, out success);
            if (!Interface.BoolOf(success))
            {
                Typesystem.LibraryExceptions.Instance.Throw();
            }
            return queueSize;
        }
        #endregion

        #region Debug

        /// <summary>
        /// Turn simulation of overflow on/off. For test purposes.
        /// <para/>
        /// Setting inQueues to true means that no messages or requests are handled by the application.
        /// An incoming request will result in an overflow, and an incoming message will be discarded.
        /// Setting outQueues to true means that no messages or requests can be sent from the application,
        /// instead these calls will throw a Safir::Dob::OverflowException. When reset to false
        /// OnXxxxNotOverflow will be called as expected.
        /// Use this to verify that your application handles overflows correctly.
        /// <para/>
        /// Note that the inQueues flag is not applied to new consumers added after this call.
        /// </summary>
        /// <param name="inQueues">If true all incoming queues are simulated full.</param>
        /// <param name="outQueues">If true all outgoing queues are simulated full.</param>
        public void SimulateOverflows(bool inQueues, bool outQueues)
        {
            byte success;
            Interface.DoseC_SimulateOverflows(ControllerId,
                                              Interface.ByteOf(inQueues),
                                              Interface.ByteOf(outQueues),
                                              out success);
            if (!Interface.BoolOf(success))
            {
                Safir.Dob.Typesystem.LibraryExceptions.Instance.Throw();
            }
        }

        #endregion

        #region Shared Memory statistics
        
        /// <summary>
        /// Get the number of currently used bytes in the shared memory.
        /// <para/>
        /// The size of the shared memory is defined by the parameter
        /// Safir.Dob.NodeParameters.SharedMemorySize, which is defined in megabytes (1024*1024 bytes).
        /// <para/>
        /// Calling this function does not require the underlying Connection to have been opened.
        /// </summary>
        /// <returns>The amount of shared memory used, in bytes.</returns>
        public System.Int64 GetSharedMemoryUsage()
        {
            System.Int64 usage;
            byte success;
            Interface.DoseC_GetSharedMemoryUsage(out usage,
                                                 out success);
            if (!Interface.BoolOf(success))
            {
                Safir.Dob.Typesystem.LibraryExceptions.Instance.Throw();
            }

            return usage;
        }

        #endregion
    }
}
